zerver/lib/actions.py

```diff
@@ -3799,11 +3799,27 @@ def do_update_message(user_profile: UserProfile, message: Message, topic_name: O
     }  # type: Dict[str, Any]
     changed_messages = [message]
 
+    send_event_users = list()  # call update event for these users
+    ums = UserMessage.objects.filter(message=message.id)
+    ums_user_ids = [user.user_profile_id for user in ums]
     if message.is_stream_message():
         stream_id = message.recipient.type_id
         event['stream_name'] = Stream.objects.get(id=stream_id).name
-
-    ums = UserMessage.objects.filter(message=message.id)
+        # Get subscribers of the stream who are active now.
+        active_subs = Subscription.objects.filter(recipient__type=Recipient.STREAM,
+                                                  recipient__type_id=stream_id,
+                                                  active=True)
+        # Also send update event to subscribers who are not
+        # not in ums, i.e. they subscribed to the stream after
+        # the message was sent, or are not subscribed at all.
+        # TODO: Send update event to active non-subscribers who
```

Comment:
I investigated a bit but couldn't figure out a good way to do this. Before this @timabbott  is this worth doing this as there isn't any good use case of this?

Reply:
And a question out curiosity, do we add UM row even if the user just read a stream without subscribing?

Reply:
No, we are not adding a UM row. A UM row not present for the user for stream's past messages is the problem here. We are simply sending the update event to the user client. We still are not doing anything for the non-subscribed users who can see a stream's history.